discovery_engine_core/core/src/storage/sqlite.rs

```diff
@@ -1129,4 +1215,60 @@ mod tests {
 
         assert!(storage.state().clear().await.unwrap());
     }
+
+    #[tokio::test]
+    async fn test_source_preference() {
+        let storage = create_memory_storage().await;
+
+        // if no sources are set, return an empty set
+        let trusted_db = storage.source_preference().fetch_trusted().await.unwrap();
+        assert!(trusted_db.is_empty());
+        let excluded_db = storage.source_preference().fetch_excluded().await.unwrap();
+        assert!(excluded_db.is_empty());
+
+        // set sources and fetch them
+        // the sources fetched should match the sources previously set
+        let trusted_sources = hashset! {"a".to_string(), "b".to_string()};
+        let excluded_sources = hashset! {"c".to_string()};
+
+        storage
+            .source_preference()
+            .set_trusted(&trusted_sources)
+            .await
+            .unwrap();
+        storage
+            .source_preference()
+            .set_excluded(&excluded_sources)
+            .await
+            .unwrap();
+
+        let trusted_db = storage.source_preference().fetch_trusted().await.unwrap();
+        assert_eq!(trusted_db, trusted_sources);
+        let excluded_db = storage.source_preference().fetch_excluded().await.unwrap();
+        assert_eq!(excluded_db, excluded_sources);
+
+        // set the excluded source and so far unknown source as new trusted sources
+        // excluded sources should be empty
+        // trusted sources should return {"d", "c"}
```

Comment:
:art:  would help to clarify a bit as in:
```suggestion
        // set the excluded source "c" and so far unknown source "d" as new trusted sources
        // excluded sources should be empty
        // trusted sources should return {"d", "c"}
```
:grey_question:  so when you set an excluded source as trusted, it stops being excluded. how about if you set a trusted source as excluded, is it worth testing this as well?

Reply:
good point 👍 i have added a test for this as well